R/parse_variable_indices.R

```diff
@@ -0,0 +1,148 @@
+#' the variable names. Each element contains: 
+#'    $dimensionality  the number of dimensions of V. Returns 0 for scalars with no brackets
+#'    but 1 for `y[1]` even if `y` has no other entries in `x`.
+#'    
+#'    $dimensions  a vector of the actual dimensions of V, as determined by the number of unique
+#'    elements at each index position. Set to `NA` if the dimensionality is zero.
+#'    
+#'    $implied_dimensions  a vector of the implied dimensions of V, where any position in V that 
+#'    contains exclusively integers is filled in to include all integers from the lesser of one 
+#'    and its minimum up to its maximum. Set to `NA` if the dimensionality is zero.
+#'    
+#'    $index_names  a list of length corresponding to the dimensionality, where each element is the
+#'    unique levels of the corresponding index if the index is parsed as factor, and NULL otherwise.
+#'    Set to `NULL` if the dimensionality is zero.
+#'    
+#'    $indices  if dimensionality is zero, returns 1.
+#'    if dimensionality is 1 or greater, returns a dataframe of every implied combination of indices
+#'    
+#'    $position  the position of each combination of indices from $indices in the the argument `x`
+#' 
+#' @details
+#' Assumes that variable indexing uses square brackets in the variable names
+#' 
+NULL
+
+parse_variable_indices <- function(x){
+  vars_indices <- strsplit(x, "(\\[|\\])")
+  vars <- sapply(vars_indices, `[[`, 1)
+  var_names <- unique(vars)
+  # Check that no variables contain unpaired or non-terminal square brackets
+  if_indexed <- (sapply(vars_indices, length) > 1)
+  if_indexed2 <- grepl("\\[.*\\]$", x)
+  extra_bracket <- grepl("\\].", x) | grepl("\\[.*\\[", x)
+  if (any(extra_bracket) | (!identical(if_indexed, if_indexed2))) {
+    stop_no_call(paste("Some variable names contain unpaired square brackets",
+                  "or are multi-indexed."))
+  }
+  missing_index <- grepl("\\[,|,\\]|,,", x) # check for `[,` `,,` or `,]`
+  if (any(missing_index)) {
+    stop_no_call(paste("Some variables contain missing indices. Each comma between square", 
+                       "brackets must be both preceeded and succeeded by an index."))
+  }
+
+  # Get dimensionality. Variables with no brackets are given as dimensionality zero.
+  # Variables with brackets are given dimensionality 1, even if they contain just one element.
+  dimensionality_elementwise <- sapply(vars_indices, function(x){
+    if (length(x) == 2) {
+      out <- length(strsplit(x[2], ",")[[1]]) # number of commas plus one
+    } else {
+      out <- 0
+    }
+    out
+  })
+  dimensionality <- sapply(var_names, function(x){
+    out <- unique(dimensionality_elementwise[vars == x])
+    if (length(out) != 1) {
+      stop_no_call(paste0("Inconsistent indexing found for variable ", x, " ."))
+    }
+    out
+  })
+  variable_indices_info <- lapply(var_names, function (x) {
```

Comment:
You have three instances of *apply over the same structure (`var_name`) in a row here. Is each element of the intermediate variables (`dimensionality[[i]]`, `dimensionality_elementwise[[i]]`) only used in corresponding element `i` of this last apply? If so I would skip creating these intermediate variables and move the relevant code into the final apply. This would also resolve my previous comment.

Reply:
I could well be missing a clever trick here, but I think this and the previous comment are not quite straightforward.

Of these three apply statements, the first applies a function over `vars_indices` and the second and third apply over `var_names`.  It would be easy to bring the second inside the third.  However, the second apply statement is very cheap and includes a check for inconsistent dimensionality.  I thought it might be better to error quickly on inconsistent dimensionality rather than start the relatively expensive variable parsing.

Reply:
Ah I think I misunderstood what this was doing. 

I think this still could be simplified so that the parsing of variable indices only has to be done once (which should be faster and also have lower maintenance cost, since we wouldn't have to modify that parsing in two places for future features). If you move [this line](https://github.com/stan-dev/posterior/pull/152/files#diff-27e20aebb5dcbf1421dc3aaa68d69b3ebc556f813dadadff19d5814951f70cd6R81):

```r
      indices <- sapply(vars_indices[var_i], `[[`, 2)
```

Above its containing if/else block, then I think you could also factor out the vectorized strsplit from the line after the above one to get a list of indices, upon which you could then use lengths / etc to get elementwise ndims and check for consistency. Then index parsing only happens once / in one place.